### PR TITLE
fix(vrl): correct conversion of datetime patterns to strftime format

### LINKED ISSUE #12185: The `http_pipelines_blackhole` soak seems to be partially broken
While investigating https://github.com/vectordotdev/vector/issues/12184 we discovered that the `http_pipelines_blackhole` sink was outputting many errors (resulting in a 1.1 GB stderr capture file). It appears like the configuration may not be completely valid. This could be affecting the soak results.

```
2022-04-04T17:13:52.134446Z  WARN transform{component_kind="transform" component_id=processing.logs.pipelines.python.truthy.transforms.0 component_type=remap component_name=processing.logs.pipelines.python.truthy.transforms.0}: datadog_grok::parse_grok: Error applying filter field=process.id filter=Integer error=failed to apply filter 'Integer' to '""'
2022-04-04T17:13:52.134524Z ERROR transform{component_kind="transform" component_id=processing.logs.pipelines.python.truthy.transforms.0 component_type=remap component_name=processing.logs.pipelines.python.truthy.transforms.0}: datadog_grok::matchers::date: Error parsing date date= error=premature end of input
2022-04-04T17:13:52.134528Z  WARN transform{component_kind="transform" component_id=processing.logs.pipelines.python.truthy.transforms.0 component_type=remap component_name=processing.logs.pipelines.python.truthy.transforms.0}: datadog_grok::parse_grok: Error applying filter field=timestamp filter=Date error=failed to apply filter 'date("yyyy-MM-dd'T'HH:mm:ss','SSS")' to ''
2022-04-04T17:13:52.135305Z  WARN transform{component_kind="transform" component_id=processing.logs.pipelines.python.truthy.transforms.0 component_type=remap component_name=processing.logs.pipelines.python.truthy.transforms.0}: datadog_grok::parse_grok: Error applying filter field=process.id filter=Integer error=failed to apply filter 'Integer' to '""'
2022-04-04T17:13:52.135383Z ERROR transform{component_kind="transform" component_id=processing.logs.pipelines.ruby.truthy.transforms.3 component_type=remap component_name=processing.logs.pipelines.ruby.truthy.transforms.3}: datadog_grok::matchers::date: Error parsing date date=1999-03-03T02:34:24.895701 error=trailing input
2022-04-04T17:13:52.135392Z  WARN transform{component_kind="transform" component_id=processing.logs.pipelines.ruby.truthy.transforms.3 component_type=remap component_name=processing.logs.pipelines.ruby.truthy.transforms.3}: datadog_grok::parse_grok: Error applying filter field=date filter=Date error=failed to apply filter 'date("yyyy-MM-dd'T'HH:mm:ss.SSSSSS")' to '1999-03-03T02:34:24.895701'
2022-04-04T17:13:52.135396Z ERROR transform{component_kind="transform" component_id=processing.logs.pipelines.ruby.truthy.transforms.3 component_type=remap component_name=processing.logs.pipelines.ruby.truthy.transforms.3}: datadog_grok::matchers::date: Error parsing date date= error=premature end of input
2022-04-04T17:13:52.135405Z  WARN transform{component_kind="transform" component_id=processing.logs.pipelines.ruby.truthy.transforms.3 component_type=remap component_name=processing.logs.pipelines.ruby.truthy.transforms.3}: datadog_grok::parse_grok: Error applying filter field=date filter=Date error=failed to apply filter 'date("yyyy-MM-dd'T'HH:mm:ss.SSS")' to ''
2022-04-04T17:13:52.135747Z ERROR transform{component_kind="transform" component_id=processing.logs.pipelines.ruby.truthy.transforms.3 component_type=remap component_name=processing.logs.pipelines.ruby.truthy.transforms.3}: datadog_grok::matchers::date: Error parsing date date=1999-03-03T02:34:24.895701 error=trailing input
2022-04-04T17:13:52.135753Z  WARN transform{component_kind="transform" component_id=processing.logs.pipelines.ruby.truthy.transforms.3 component_type=remap component_name=processing.logs.pipelines.ruby.truthy.transforms.3}: datadog_grok::parse_grok: Error applying filter field=date filter=Date error=failed to apply filter 'date("yyyy-MM-dd'T'HH:mm:ss.SSSSSS")' to '1999-03-03T02:34:24.895701'
2022-04-04T17:13:52.135756Z ERROR transform{component_kind="transform" component_id=processing.logs.pipelines.ruby.truthy.transforms.3 component_type=remap component_name=processing.logs.pipelines.ruby.truthy.transforms.3}: datadog_grok::matchers::date: Error parsing date date= error=premature end of input
2022-04-04T17:13:52.135759Z  WARN transform{component_kind="transform" component_id=processing.logs.pipelines.ruby.truthy.transforms.3 component_type=remap component_name=processing.logs.pipelines.ruby.truthy.transforms.3}: datadog_grok::parse_grok: Error applying filter field=date filter=Date error=failed to apply filter 'date("yyyy-MM-dd'T'HH:mm:ss.SSS")' to ''
2022-04-04T17:13:52.135846Z  WARN transform{component_kind="transform" component_id=processing.logs.pipelines.nginx.truthy.transforms.0 component_type=remap component_name=processing.logs.pipelines.nginx.truthy.transforms.0}: datadog_grok::parse_grok: Error applying filter field=duration filter=Number error=failed to apply filter 'Number' to '""'
2022-04-04T17:13:52.135865Z ERROR transform{component_kind="transform" component_id=processing.logs.pipelines.ruby.truthy.transforms.3 component_type=remap component_name=processing.logs.pipelines.ruby.truthy.transforms.3}: datadog_grok::matchers::date: Error parsing date date=1999-03-03T02:34:24.895701 error=trailing input
2022-04-04T17:13:52.135869Z  WARN transform{component_kind="transform" component_id=processing.logs.pipelines.ruby.truthy.transforms.3 component_type=remap component_name=processing.logs.pipelines.ruby.truthy.transforms.3}: datadog_grok::parse_grok: Error applying filter field=date filter=Date error=failed to apply filter 'date("yyyy-MM-dd'T'HH:mm:ss.SSSSSS")' to '1999-03-03T02:34:24.895701'
2022-04-04T17:13:52.135872Z ERROR transform{component_kind="transform" component_id=processing.logs.pipelines.ruby.truthy.transforms.3 component_type=remap component_name=processing.logs.pipelines.ruby.truthy.transforms.3}: datadog_grok::matchers::date: Error parsing date date= error=premature end of input
2022-04-04T17:13:52.135874Z  WARN transform{component_kind="transform" component_id=processing.logs.pipelines.ruby.truthy.transforms.3 component_type=remap component_name=processing.logs.pipelines.ruby.truthy.transforms.3}: datadog_grok::parse_grok: Error applying filter field=date filter=Date error=failed to apply filter 'date("yyyy-MM-dd'T'HH:mm:ss.SSS")' to ''
2022-04-04T17:13:52.135874Z  WARN transform{component_kind="transform" component_id=processing.logs.pipelines.nginx.truthy.transforms.0 component_type=remap component_name=processing.logs.pipelines.nginx.truthy.transforms.0}: datadog_grok::parse_grok: Error applying filter field=duration filter=Number error=failed to apply filter 'Number' to '""'
2022-04-04T17:13:52.135978Z ERROR transform{component_kind="transform" component_id=processing.logs.pipelines.ruby.truthy.transforms.3 component_type=remap component_name=processing.logs.pipelines.ruby.truthy.transforms.3}: datadog_grok::matchers::date: Error parsing date date=2018-06-27T08:01:52.991728 error=trailing input
2022-04-04T17:13:52.135985Z  WARN transform{component_kind="transform" component_id=processing.logs.pipelines.ruby.truthy.transforms.3 component_type=remap component_name=processing.logs.pipelines.ruby.truthy.transforms.3}: datadog_grok::parse_grok: Error applying filter field=date filter=Date error=failed to apply filter 'date("yyyy-MM-dd'T'HH:mm:ss.SSSSSS")' to '2018-06-27T08:01:52.991728'
2022-04-04T17:13:52.135987Z ERROR transform{component_kind="transform" component_id=processing.logs.pipelines.ruby.truthy.transforms.3 component_type=remap component_name=processing.logs.pipelines.ruby.truthy.transforms.3}: datadog_grok::matchers::date: Error parsing date date= error=premature end of input
2022-04-04T17:13:52.135990Z  WARN transform{component_kind="transform" component_id=processing.logs.pipelines.ruby.truthy.transforms.3 component_type=remap component_name=processing.logs.pipelines.ruby.truthy.transforms.3}: datadog_grok::parse_grok: Error applying filter field=date filter=Date error=failed to apply filter 'date("yyyy-MM-dd'T'HH:mm:ss.SSS")' to ''
```

We should check / update the other pipelines soaks too.

### PR DESCRIPTION
**Pull request by vladimir-dd**
Closes #12185

Fixes a couple of issues with `date` filter of `parse_groks` function, discovered during soak testing.

### CODE REVIEW

File: lib/datadog/grok/src/matchers/date.rs [hunk 1]
```diff
@@ -219,7 +227,13 @@ pub fn time_format_to_regex(
 pub fn apply_date_filter(value: &Value, filter: &DateFilter) -> Result<Value, GrokRuntimeError> {
     match value {
         Value::Bytes(bytes) => {
-            let value = String::from_utf8_lossy(bytes);
+            let mut value = String::from_utf8_lossy(bytes).to_string();
```
Comment (blt) [inline lib/datadog/grok/src/matchers/date.rs:230]:
```suggestion
            let mut value = String::from_utf8_lossy(bytes).into_owned();
```

Relying on the method on `Cow<'_, str>` makes this more clear to my eye.

**Review by blt — APPROVED** [review]
One small suggestion, not blocking. Otherwise looks good to me.

File: lib/datadog/grok/src/matchers/date.rs [hunk 2]
```diff
@@ -219,7 +227,13 @@ pub fn time_format_to_regex(
...
+            // ideally this Z should in quoted in the pattern, but DataDog supports this as a special case:
```
Comment (StephenWakely) [inline lib/datadog/grok/src/matchers/date.rs:231]:
```suggestion
            // Ideally this Z should be quoted in the pattern, but DataDog supports this as a special case:
```

File: lib/datadog/grok/src/matchers/date.rs [hunk 3]
```diff
@@ -24,8 +24,16 @@ pub fn convert_time_format(format: &str) -> std::result::Result<String, String>
                 // second of minute
                 's' => time_format.push_str("%S"),
                 // fraction of second
-                'S' => time_format.push_str("%3f"),
+                'S' => {
+                    if time_format.ends_with('.') {
+                        time_format.remove(time_format.len() - 1); // drop .
```
Comment (StephenWakely) [inline lib/datadog/grok/src/matchers/date.rs:29]: You can use [pop](https://doc.rust-lang.org/std/string/struct.String.html#method.pop) to drop the last character. I think that would be clearer.

File: lib/datadog/grok/src/matchers/date.rs [hunk 4]
```diff
@@ -219,7 +227,13 @@ pub fn time_format_to_regex(
 pub fn apply_date_filter(value: &Value, filter: &DateFilter) -> Result<Value, GrokRuntimeError> {
     match value {
         Value::Bytes(bytes) => {
-            let value = String::from_utf8_lossy(bytes);
+            let mut value = String::from_utf8_lossy(bytes).to_string();
+            // ideally this Z should in quoted in the pattern, but DataDog supports this as a special case:
+            // yyyy-MM-dd'T'HH:mm:ss.SSSZ - e.g. 2016-09-02T15:02:29.648Z
+            if value.ends_with('Z') && filter.original_format.ends_with('Z') {
+                value.remove(value.len() - 1); // drop Z
```
Comment (StephenWakely) [inline lib/datadog/grok/src/matchers/date.rs:234]: Likewise here with [pop](https://doc.rust-lang.org/std/string/struct.String.html#method.pop).

**Review by StephenWakely — APPROVED** [review]
Looks good.

File: lib/datadog/grok/src/matchers/date.rs (hunk 3 above)
Comment (vladimir-dd) [reply lib/datadog/grok/src/matchers/date.rs]: definitely, thanks!

File: lib/datadog/grok/src/matchers/date.rs [hunk 5]
```diff
@@ -234,12 +248,12 @@ pub fn apply_date_filter(value: &Value, filter: &DateFilter) -> Result<Value, Gr
                         .expect("this regex should always contain tz group")
                         .as_str();
                     let tz: Tz = tz.parse().map_err(|error| {
-                        error!(message = "Error parsing tz", tz = %tz, % error);
+                        warn!(message = "Error parsing tz", tz = %tz, % error);
```
[the remark this answers is not in the document]
Comment (vladimir-dd) [reply lib/datadog/grok/src/matchers/date.rs:251]: So this line can be hit at runtime when we fail to recognize a timezone. This is a fair question, in this and other places I log warns, but wondering what would be the best strategy here.. These logs could help users understand if their grok patterns fail to parse logs at they expect. @jszwedko If you agree, I'd add rate-limiting to all these logs in a separate PR.

[the remark this answers is not in the document]
Comment (jszwedko) [reply lib/datadog/grok/src/matchers/date.rs:251]: I think we should just bubble this up as an error that is returned to `parse_grok`. The user could then decide what they want to do with it (e.g. log it with the `log()` function in VRL).

**Pull request merged**